torch/csrc/jit/graph_executor.cpp

```diff
@@ -305,137 +329,153 @@ struct GraphExecutorImpl {
+    FuseGraph(graph);
+  }
+
+  static bool needsGradient(const std::shared_ptr<const Graph>& graph, const ArgumentSpec& spec) {
+    if (!autograd::GradMode::is_enabled())
+      return false;
+    if (mayIntroduceGradient(graph->block()))
+      return true;
     for(size_t i = 0; i < spec.size(); ++i) {
       if(spec.at(i).requires_grad())
         return true;
     }
     return false;
   }
 
-  ExecutionPlan compileSpec(const ArgumentSpec & spec) {
-    auto graph_ = graph->copy();
+  static bool mayIntroduceGradient(const Block* b) {
+    for (const Node* n : b->nodes()) {
+      if (n->kind() == prim::PythonOp)
+        return true;
+      for (const Block* bb : n->blocks()) {
+        if (mayIntroduceGradient(bb))
+          return true;
+      }
+    }
+    return false;
+  }
 
-    specializeToSpec(graph_, spec);
+  void runTraced(Stack & stack) {
+    auto state = tracer::getTracingState();
+    auto inputs = last(stack, num_inputs);
+    auto input_values = fmap(inputs, [](const IValue & v) {
+      return tracer::getValueTrace(v.toTensor());
+    });
 
-    if(!argumentSpecRequiresGradient(spec)) {
-      runOptimization(graph_, /*graphMustSupportVariables=*/false);
-      return ExecutionPlan(graph_);
-    }
-    JIT_ASSERT(symbolically_differentiable);
+    ArgumentSpec spec(autograd::GradMode::is_enabled(), inputs);
+    // NB: we could just run the fallback in here and call it a day, but that would loose all
+    // the control flow information we have in the graph. Thus, we run the fallback to
+    // get the correct output values, but we will override the tracing states later.
+    getOrCompileFallback().run(stack);
 
-    std::vector<bool> requires_grads;
-    requires_grads.reserve(spec.size());
-    for(size_t i = 0; i < spec.size(); i++)
-      requires_grads.push_back(spec.at(i).requires_grad());
+    // Traces always have types propagated through them, so we make sure to
+    // also propagate types through the graph we are inserting here.
+    // However, this->graph itself may already have been generated with
+    // tracing and so we only do the type propgation if no concrete types have
+    // been set.
+    auto local_graph = this->graph->copy();
+    PropagateInputShapes(*local_graph, spec);
+    auto output_values = script::inlineCallTo(*state->graph, *local_graph, input_values);
 
-    Gradient gradient = differentiate(graph_, requires_grads);
-    graph_ = gradient.f;
-    runOptimization(graph_, /*graphMustSupportVariables=*/false);
-    return ExecutionPlan(graph_, std::move(gradient));
+    auto outputs = last(stack, num_outputs);
+    for (size_t i = 0; i < outputs.size(); ++i) {
+      // We can't attach tracing states to scalars, so we have to skip them here
+      // TODO: Should we reinterpret them as scalar tensors instead?
+      if (!outputs[i].isTensor()) continue;
+      tracer::setValueTrace(outputs[i].toTensor(), output_values[i]);
+    }
   }
+  // TODO TODO TODO TODO update comments here here
```

Comment:
yes, especially the now-inaccurate stuff below about how the fallback used to work.

Reply:
Oh yes. I put 4 TODOs to make sure it's hard to miss during a review in case I forgot 😄